Adds Dropdowns to the Navbar

joshwingreene opened the pull request:
## Changes
- Adds a `NavCategory` component. Whenever a developer wants to add a dropdown for a major navigation area, they should make use of this component. Links are added through the use of the `CustomNavLink` component. Only one NavCategory can be expanded at a time.
- Moves the `NavLinkConfig` type to the `CustomNavLink` component

## Purpose
- Adds a NavDropdown component that can be used to split up the main navigation in the navbar with dropdowns

## Approach
- Based on @DropsOfSerenity's recommendation, I based this implementation on how the FilterCategory is used by the FilterDropdown for the DataTable.

## Testing Steps
1. Pull in the changes to your local copy of this branch and run it alongside the dj_starter_demo repo.

## Screenshots

![dropdown-pr-update-1](https://user-images.githubusercontent.com/2876874/187782215-20453b9e-cb8b-4ba1-8eed-f7ebcf7c0b76.png)

![dropdown-pr-update-2](https://user-images.githubusercontent.com/2876874/187782235-1b37dde9-b417-42d9-b92f-e4a26843f1da.png)

![dropdown-pr-update-3](https://user-images.githubusercontent.com/2876874/187782283-885038fc-6683-4f36-b944-be6464697e7d.png)

![dropdown-pr-update-4](https://user-images.githubusercontent.com/2876874/187782344-2bb676b0-1467-49b9-99b8-5eb1e794463d.png)

joshwingreene committed:
feat(Dropdown and VerticalNav): Started to construct the Dropdown component

joshwingreene committed:
feat(various): Started to implement an approach that's based on how the FilterDropdown works

joshwingreene committed:
feat(NavDropdown and VerticalNav): Implemented the component.

joshwingreene committed:
feat(various): Finished implementation. Just need to refactor the names of the styled components.

joshwingreene committed:
refactor(NavDropdown and VerticalNav): Renamed props, parameters, and the FilterCategory component that was copied over from the FilterDropdown component

Also, edited the padding of the content area.

DropsOfSerenity commented on src/features/navbar/components/VerticalNav.tsx:
```diff
@@ -38,6 +41,21 @@ export const VerticalNav: FC<Props> = ({ closeVerticalNav }) => {
 
   const defaultLanguageOption = __languageOptions.find(language => language.value === i18n.languages[0]);
 
+  const navCategoryMap = {
+    General: navLinks.filter(link => link.label === 'Directory'),
+    Management: navLinks.filter(link => link.label === 'Users'),
+  };
+
+  const getDefaultCategoryBasedOnLocation = () => {
+    if (location.pathname === '/users') {
+      return 'Management';
+    }
+    if (location.pathname === '/agents') {
+      return 'General';
+    }
+    return null;
+  };
```
This is not very maintainable. Whenever someone adds a new navlink they need to reach in here and make sure to update these accordingly. The `navCategoryMap` in particular would be difficult to keep updated since there is no direct relationship between a category and a navLink.

Instead what do you think of removing the abstraction of `navLinks` entirely. Lets just declarative define them in the JSX.

An API like this looks like it would solve all the problems, while being easy to maintain.

![image](https://user-images.githubusercontent.com/2287977/187499064-b5c7296c-1b9c-4bee-aab6-7339b1c25960.png)

joshwingreene replied:
Yeah. I agree. And, replacing the navLinks abstraction with a declarative JSX approach makes sense to me. I'll go ahead and refactor the current implementation. 

Also, thanks for the API Design tip. I'll keep it in mind for future API design decisions.

DropsOfSerenity reviewed (changes requested):
The visuals are good but I think we can make this simpler.

One of the things I find helpful in designing API's is writing the usage code first and seeing if you like it.

joshwingreene committed:
refactor(various): Improved the API for the Navbar Dropdown implementation

Also, removed the useNavLinks hook and moved the NavLinkConfig type to the CustomNavLink component.

joshwingreene committed:
refactor(various): Only one category can be opened at a time; Simplifies the category api

joshwingreene committed:
fix(NavCategory): Forgot some code that was supposed to be in the last commit

joshwingreene committed:
feat(various): A category will be open by default if the user navigates to one of its associated paths.

DropsOfSerenity commented:
Since navbar is moving to topbar in #634 , I'm closing this for now.
